### PR TITLE
Input Selection Error type

### PR DESCRIPTION
**Pull request by thibault-martinez**
## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: client/src/error.rs [hunk 1]
```diff
@@ -205,47 +201,9 @@ pub enum Error {
...
-        found: U256,
-        /// The required amount.
-        required: U256,
-    },
-    /// Can't burn and transition an output at the same time.
-    #[error("can't burn and transition an output at the same time, chain ID: {0}")]
-    BurnAndTransition(ChainId),
+    /// Input selection error.
+    #[error("{0}")]
+    InputSelection(#[from] InputSelectionError),
```
Comment (Thoralf-M) [inline client/src/error.rs:192]: What about the following variants? They are not in the ISA core, but still related to input selection I would say

```Rust
    /// The wallet account has enough funds, but split on too many outputs
    #[error("the wallet account has enough funds, but split on too many outputs: {0}, max. is 128, consolidate them")]
    ConsolidationRequired(usize),
    /// Missing input for utxo chain
    #[error("missing input: {0}")]
    MissingInput(String),
    /// No input with matching ed25519 address provided
    #[error("no input with matching ed25519 address provided")]
    MissingInputWithEd25519Address,
    /// The wallet account doesn't have any inputs found
    #[error("no inputs found")]
    NoInputs,
    /// The wallet account doesn't have enough native tokens
    #[error("the wallet account doesn't have enough native tokens, missing: {0:?}")]
    NotEnoughNativeTokens(NativeTokens),
    /// The wallet account doesn't have enough balance for an output with the remaining native tokens.
    #[error("the wallet account doesn't have enough balance for an output with the remaining native tokens.")]
    NoBalanceForNativeTokenRemainder,
```

Some can also be removed or changed I think, for example `ConsolidationRequired` is more like `TooManyInputs`, `NotEnoughNativeTokens` is basically the same as `InsufficientNativeTokenAmount` and `NoBalanceForNativeTokenRemainder` is just a clearer name for `InsufficientAmount`
`MissingInputWithEd25519Address` is in the new Error, but not removed here

Comment (thibault-martinez) [reply client/src/error.rs:192]: `MissingInput` was not used anymore so I removed it.
`MissingInputWithEd25519Address` removed from the main error type.
`NoInputs` replaced by existing `NoAvailableInputsProvided`.
`NoBalanceForNativeTokenRemainder` wasn't used anymore, removed.
`NotEnoughNativeTokens` wasn't used anymore, removed.
`ConsolidationRequired` I personally wouldn't put in ISA tbh, or later when I'm done with other error refactoring see where it best fits.

File: client/src/api/block_builder/input_selection/automatic.rs [hunk 2]
```diff
@@ -218,33 +218,19 @@ impl<'a> ClientBlockBuilder<'a> {
...
+                        Err(InputSelectionError::Block(block_error)) => match block_error {
                             iota_types::block::Error::InvalidStorageDepositAmount { .. } => {
-                                cached_error.replace(crate::Error::Block(block_error));
+                                cached_error.replace(Error::from(InputSelectionError::Block(block_error)));
                                 continue;
                             }
                             _ => return Err(block_error.into()),
                         },
-                        Err(e) => return Err(e),
+                        Err(e) => return Err(e)?,
```
Comment (Thoralf-M) [inline client/src/api/block_builder/input_selection/automatic.rs]: InsufficientAmount, InsufficientNativeTokenAmount and NoAvailableInputsProvided should still not return an error here?

Comment (thibault-martinez) [reply client/src/api/block_builder/input_selection/automatic.rs]: Added back

Comment (Thoralf-M) [reply client/src/api/block_builder/input_selection/automatic.rs]: Can we event get `InsufficientAmount` there? And what about `NoAvailableInputsProvided`?

File: client/src/api/block_builder/input_selection/core/error.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2021 IOTA Stiftung
```
Comment (Thoralf-M) [inline client/src/api/block_builder/input_selection/core/error.rs]:
```suggestion
// Copyright 2023 IOTA Stiftung
```

File: client/src/api/block_builder/input_selection/core/error.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+use std::fmt::Debug;
+
+use iota_types::block::output::{ChainId, OutputId, TokenId};
+use primitive_types::U256;
+use serde::Serialize;
+
+use crate::{api::input_selection::Requirement, error::display_string};
+
+/// Error type of the iota client crate.
```
Comment (Thoralf-M) [inline client/src/api/block_builder/input_selection/core/error.rs]: Comments should probably mention ISA

Comment (thibault-martinez) [reply client/src/api/block_builder/input_selection/core/error.rs]: Indeed

Comment (Thoralf-M) [reply client/src/api/block_builder/input_selection/core/error.rs]: And the `//! Error handling in iota-client crate.`?

Comment (thibault-martinez) [reply client/src/api/block_builder/input_selection/core/error.rs]: Oh ffs

Comment (thibault-martinez) [reply client/src/api/block_builder/input_selection/core/error.rs]: Done

**Review by Thoralf-M — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Pull request merged**